### ISSUE TITLE
Volumetric (3D + channels) Affine warping

### ISSUE THREAD

**Issue by mibaumgartner** [kornia/kornia#563]
## 🚀 Feature Request
Add support for volumetric data (C, (D,) H, W) and affine transformations

## Motivation
Volumetric transformations can be used for any kind of volumetric data (e.g. medical -> CT, MRI) and affine transformations are very common for augmentation transformations (e.g. in [rising](https://github.com/PhoenixDL/rising)).

## Pitch
Introduce a new base class `GenericWarper` from which `DepthWarper` , `HomographyWarper` and `AffineWarper` can be derived from.

The functional API of the warp functions shouldn't be changed because it can automatically detect if the input tensor has an additional dimension or not.

I'm not quite sure how the volumetric support should be added for the future. One possibility would be to add a first version exists besides the 2D functions and has support for both 2D and 3D (could probably look similar to rising's functionals) or add additional functions to specifically/only handle 3D data (this would probably lead to quite a lot of code duplication).

## Additional context

@justusschock

**Comment by edgarriba** [comment]
my idea was to refactor `kornia.geometry.warp` with something like this.
/cc @ducha-aiki @lferraz had some initial thoughts as well since he refactored few weeks ago to handle cases where you want to reuse the grid to speed up processes.

```python
class BaseGridSampleWarper(nn.Module):
  """Encapsulates torch.grid_sample"""

  def compute_grid(self, transform: torch.Tensor) -> torch.Tensor:
    raise NotImplementedErrror("Overrride depending on your transform type)

  def forward(self, input: torch.Tensor, transform: torch.Tensor) -> torch.Tensor:
    grid: torch.Tensor = self.compute_grid(transform)
    assert grid.shape == (B, H, W, 2)
    return torch.nn.fuctional.grid_sample(input, grid, ....)
```
Then any type of transformation can be specialized from this by just adjusting the way it produces the grid depending on the desired behavior.
```python
class AffineWarper(BaseGridSampleWarper):
  r"""Warp a tensor using a rigid 2d affine transformation."""
  def compute_grid(self, transform: torch.Tensor) -> torch.Tensor:
    # from https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/imgwarp.py#L147
    B, C, H, W = src.size()
    dsize_src = (H, W)
    out_size = dsize
    # we generate a 3x3 transformation matrix from 2x3 affine
    M_3x3: torch.Tensor = convert_affinematrix_to_homography(M)
    dst_norm_trans_src_norm: torch.Tensor = normalize_homography(
        M_3x3, dsize_src, out_size)
    src_norm_trans_dst_norm = torch.inverse(dst_norm_trans_src_norm)
    grid = F.affine_grid(src_norm_trans_dst_norm[:, :2, :],  # type: ignore
                         [B, C, out_size[0], out_size[1]],
                         align_corners=align_corners)
    return grid
```
```python
class PerspectiveWarper(BaseGridSampleWarper):
  r"""Warp a tensor using a rigid 2d perspective transformation."""
  def compute_grid(self, transform: torch.Tensor) -> torch.Tensor:
    # we need to double check we can produce same with affine_grid
    return grid
```
```python
class ProjectiveWarper(BaseGridSampleWarper):
  r"""Warp a tensor using a rigid 3d homogeneous transformation."""
  def compute_grid(self, transform: torch.Tensor) -> torch.Tensor:
    # use code from rising
    return grid
```
Then to follow OpenCV convention the high level interfaces would be:
```python
out = K.warp_affine(input, mat2x3, (H, W))  # (BxCxHxW / Bx2x3) -> BxCxHxW
out = K.warp_perspective(input, mat3x3, (H, W))  # (BxCxHxW / Bx3x3) -> BxCxHxW
out = K.warp_projective(input, mat4x4, (D, H, W))  # (BxCxDxHxW / Bx4x4) -> BxCxDxHxW
```

**Comment by kuynzereb** [comment]
Recently I was trying to torchscript `warp_perspective()` function and it turned out to be a non-trivial without significant refactoring which I have done in #574. The main problem is that all the implementation was done it `nn.Module` and functional api just creates an instance of this module and calls its methods, whereas torchscript doesn't support the dynamic creation of `nn.Module` (at least now). As far as I understand, it is better to first implement functional api and make `nn.Module`s to be just wrappers around these functions (I guess it is the way how it is done in pytorch). Then it should be quite easy to get jit versions of the functions. 

So, if someone starts to do this refactoring, keep it in mind. And actually it is not related only to these warping functions, but to any function in kornia we want to be torchscriptable.

**Comment by lferraz** [comment]
Good to know!!! Thank you so much for your contributions. I will review the code tonight :)

Luis

**Comment by edgarriba** [comment]
@kuynzereb I'm not quite sure about how torchscript handles classes since for the examples in jit seems that is doable. However, I'm in favor of splitting to simple functional since in the end everything will be more flexible.

Based on that, we could have a set of functionals that just compute the flow grids,
`create_affine_grid` `create_homography_grid` `create_perspective_grid` and future `create_undistortion_grid` , etc. All this functionals can be the ones responsible for receiving the transformation parameters do whatever normalization is needed to fulfill `grid_sample` -1/1 convention and apply [`remap`](https://kornia.readthedocs.io/en/latest/geometry.transform.html?highlight=remap#kornia.geometry.transform.remap) which basically encapsulates `grid_sample`. The current implementation for `remap` assumes non normalized coordinates, so we might have to extend the functionality to perform such normalization.

**Comment by edgarriba** [comment]
@mibaumgartner I have internally implemented already the `warp_projective` functionality plus few utlils like `get_projection_matrix` which is an extension of `get_rotation_matrix2d` that basically does the logics to compute the transformation matrix given an angle vector in angle-axis format and the 3d coordinate to the origin from where you want to apply the transformation. Besides, I needed to implement some normalization tricks to compute the final transformation is we do for the 2d affine case in [here](https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/imgwarp.py#L151-L154) but for the 3D case. Hope I can send a PR soon about it.

**Comment by edgarriba** [comment]
closing this since this feature has been merged in v0.4. @mibaumgartner please provide usage feedback on that
